### PR TITLE
DDF clone and fixes for Tuya Smart Knobs (_TZ3000_* TS004F)

### PR DESCRIPTION
**Pull request by chishm**
This follows on from the work of @manup in #6242, to add device clones of the Tuya Smart Knob and to extend the current capabilities of the DDF.

Each way of pressing and turning the knob is now indicated by a different button code. There are 2 modes ("command" and "event") and two distinct rotations in command mode (just rotate vs press down and rotate). E.g. rotating clockwise in command mode will give button code 1030, pressing and rotating in command mode gives code 2030, and rotating in event mode gives code 3030 (there is no press and rotate for event mode).

There is now a resources item (`config/mode`) to discover the current mode (`"command"` or `"event"`).

The `state/angle` value is now a relative angle for the current rotation event. It didn't make sense to have a single absolute angle when the device can be twisted in multiple ways.

The device will  now send commands to the correct group address. This is achieved using the *Add Group* command in the *Group* cluster. Having to do so is a quirk of this device (and possibly other Tuya devices).

**NB:** The group add command is performed using a script in the DDF. To support this, the C++ code was modified to allocate "auto" groups before calling any write function in the DDF.

Extending the implementation detail of @manup's work, the `state/ct` item is used to parse the *Color Control* cluster for pressed-down rotations. The item is set to `"public": false` and is not exposed to the REST-API.

**TODO:** The battery level is not being reported for some reason.

A lot of this work drew upon knowledge provided by others. I'd like to thank:
* @EduardoHerreraM for the [original DDF for the Moes Smart Knob device](https://github.com/dresden-elektronik/deconz-rest-plugin/issues/6759) that I have;
* @vvbogdanov87, @tomky, @jeroen85, and @Smanar for the [DDF for the similar TS004F _TZ3000_4fjiwweb device](https://github.com/dresden-elektronik/deconz-rest-plugin/issues/6031);
* @manup for the [PR adding support to the deCONZ](https://github.com/dresden-elektronik/deconz-rest-plugin/pull/6242) for the above, and for writing the DDF and JavaScript files that I've modified; and
* @MattWestb for the [ZHA code supporting the Tuya Smart Knobs](https://github.com/zigpy/zha-device-handlers/blob/156aaa6bb704a13fe9500b833c293405fc5ffd79/zhaquirks/tuya/ts004f.py#L68), which helpfully documents many more clones.

### CODE REVIEW

**Comment by Smanar** [comment]
>TODO: The battery level is not being reported for some reason.

Yeah, good question, I don't see the problem too, still nothing after a time ? if the binding don't work, you can have the poll. If you try to read the attribute in deconz it update the value in the API (take care you can test that only 1 time ^^)

```
    {
      "bind": "unicast",
      "src.ep": 1,
      "dst.ep": 1,
      "cl": "0x0000"
    },
```

You are sure this part is usefull ?

**Comment by manup** [comment]
>Nice, so much work on thoses device, I realy like your JS scripts with all comments, but I have a doubt, how the JS core will handle them, it will ignore them, or use memory for nothing ?

Currently the comments are read and "processed" by DucktapeJS, but that can be changed in future so they are ignored when loading the JS files and not even processed then compiling the Javascript byte code.

So imho I think it's fine to keep them, especially if they help other devs to understand the script.

**Comment by chishm** [comment]
> Nice, so much work on thoses device, I realy like your JS scripts with all comments, but I have a doubt, how the JS core will handle them, it will ignore them, or use memory for nothing ? On my side I make shorter code possible, but less instructive/reusable than yours.
> 

Thank you. I found the comments necessary for my own understanding. With all the magic constants used it was becoming hard to follow what I'd written a few days earlier. As @manup said, it'd be better if the comments could remain but be ignored during loading of the scripts.

> ```
>     {
>       "bind": "unicast",
>       "src.ep": 1,
>       "dst.ep": 1,
>       "cl": "0x0000"
>     },
> ```
> 
> You are sure this part is usefull ?

No, it's not necessary. I'd left it in after trying to debug a problem I had with retrieving the firmware version number (turned out I'd forgotten to include `tuya_swversion.js` in my DDF). I've removed it now, tested, and everything seems fine.

> > TODO: The battery level is not being reported for some reason.
> 
> Yeah, good question, I don't see the problem too, still nothing after a time ? if the binding don't work, you can have the poll. If you try to read the attribute in deconz it update the value in the API (take care you can test that only 1 time ^^)

> BTW the bot that validate DDF need to take some holidays, embittered bot ....

Turns out the bot was being helpful after all. It'd put `" eval"` instead of `"eval"` for the parse script of `config/battery`, and this was causing my battery reporting problem. I've also removed the polling, as the device reports its battery level correctly.

**Comment by Smanar** [comment]
From the bot you miss 

```
        {
          "name": "cap/color/ct/max"
        },
        {
          "name": "cap/color/ct/min"
        },
```

But here I don't understand the problem, on my side I will remove all this part.

`Error: Unrecognized key(s) in object: 'buttons', 'buttonevents'`

**Comment by Zehir** [comment]
> `Error: Unrecognized key(s) in object: 'buttons', 'buttonevents'`

It's because thoses are not valid as top level property of the DDF

**Comment by chishm** [comment]
Thanks, @Smanar and @Zehir for your help with getting the DDF syntax right. I couldn't find `devcap1.schema.json` in the repo, so I've been using existing DDF files as examples, and obviously got a few things wrong there.

**Comment by chishm** [comment]
Hi @Smanar and @manup,

Is there anything else I need to do before this can be merged?

**Comment by Smanar** [comment]
For me there is no problem, and you have the device to make test, perhaps the c++ part need a special Milestone ?

**Comment by manup** [comment]
Good work :+1: I think the PR is fine now. One question since I don't understand every detail of it, is this a breaking change REST-API wise how the modified devices work?

**Comment by chishm** [comment]
@Smanar: That makes sense that the maintainers might be waiting for the next release to merge this.

@manup: It is a breaking change for two items: the `state/angle` value is now relative, and `state/buttonevent` has some more values reported for the different ways that the knob can be pressed and turned.

**Comment by manup** [comment]
Thanks for clarification I think the change is fine in order to make the device more usable.
